tools/autograd/templates/variable_factories.h

```diff
@@ -104,117 +194,164 @@ AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
-          at::NoGradGuard guard;
-          tensor[index].fill_(elem.scalar());
-        } else if (elem.type_ == InitListTensorType::InitList) {
-          elem.fill_tensor(tensor[index]);
-        } else {
-          TORCH_INTERNAL_ASSERT(false, "Invalid InitListTensor");
-        }
+        elem.fill_tensor(tensor[index]);
         index++;
       }
+    } else if (is_tensor()) {
+      TORCH_INTERNAL_ASSERT(
+        false,
+        "TensorDataContainer is already a Tensor type, `fill_tensor` should not be called");
+    } else {
+      TORCH_INTERNAL_ASSERT(false, "Invalid TensorDataContainer type");
     }
-    c10::Scalar scalar_;
-    std::initializer_list<InitListTensor> init_list_;
-    std::vector<int64_t> sizes_;
-    c10::ScalarType scalar_type_;
-    InitListTensorType type_;
-  };
-
-  inline std::ostream& operator<<(std::ostream& stream, const InitListTensor& init_list_tensor) {
-    init_list_tensor.pretty_print_recursive(stream);
-    return stream;
   }
-} // namespace detail
 
-#define TENSOR(T, S)                                                       \
-  inline at::Tensor tensor(                                                \
-      at::ArrayRef<T> values, const at::TensorOptions& options) {          \
-    at::Tensor result = ([&]() {                                           \
-      at::AutoNonVariableTypeMode non_var_type_mode(true);                 \
-      return at::tensor(values, at::TensorOptions(options).is_variable(false)); \
-    })();                                                                  \
-    return autograd::make_variable(result, options.requires_grad());       \
-  }                                                                        \
-  inline at::Tensor tensor(                                                \
-      std::initializer_list<T> values, const at::TensorOptions& options) { \
-    return torch::tensor(at::ArrayRef<T>(values), options);                \
-  }                                                                        \
-  inline at::Tensor tensor(T value, const at::TensorOptions& options) {    \
-    return torch::tensor(at::ArrayRef<T>(value), options);                 \
-  }                                                                        \
-  inline at::Tensor tensor(at::ArrayRef<T> values) {                       \
-    return torch::tensor(std::move(values), at::dtype(at::k##S));          \
-  }                                                                        \
-  inline at::Tensor tensor(std::initializer_list<T> values) {              \
-    return torch::tensor(at::ArrayRef<T>(values));                         \
-  }                                                                        \
-  inline at::Tensor tensor(T value) {                                      \
-    return torch::tensor(at::ArrayRef<T>(value));                          \
+  std::vector<int64_t> sizes_;
+  c10::ScalarType scalar_type_;
```

Comment:
Have you considered having TensorDataContainer hold `std::variant<Scalar, Tensor, initializer_list<TensorDataContainer<D+1>>>` in this implementation ?  Would it make the implementation look nicer, worse, or roughly the same?

Reply:
https://en.cppreference.com/w/cpp/utility/variant says "A variant is not permitted to hold references, arrays, or the type void." and I guess initializer_list is a type of array?

Reply:
I am a little worried that accessing the member value of `c10::variant` would be slower than accessing it as a direct class member. Because we want to prevent perf regression on the existing `torch::tensor(at::ArrayRef<T>)` and `torch::tensor(std::vector<T>)` as much as possible, I feel that it might be better to store them as direct class members.